### ISSUE TITLE
Caret ignored when installing package and writing version to package.json

### ISSUE THREAD

**Issue by bencresty** [pnpm/pnpm#4940]
### Code to reproduce the issue:

PNPM doesn't include the caret (`^`) in `package.json`:

``` text
pnpm install gsap@^3.10.4
```

Result in package.json after using PNPM --> is missing the very important caret!!
![image](https://user-images.githubusercontent.com/69322733/175925891-cfdd6b30-d614-4c3c-a3d4-4eaab3c6d089.png)

NPM on the other hand does just work fine and is working as expected; includes the caret in the `package.json`:

``` text
npm install gsap@^3.10.4
```

Result in package.json after using NPM:
![image](https://user-images.githubusercontent.com/69322733/175925721-2e86f029-f97b-40d9-9a8f-357d3160047c.png)

### Expected behavior:

After installing with using a caret (`^`) pnpm should include this caret in the `package.json`-requirement. Because it is missing now running `pnpm update` is not working as supposed to. We included the caret intentional to keep updating 'till the next major version when running `pnpm update`. It should not be ignored when writing the requirement to `package.json`.

### Actual behavior: 

Caret gets ignored when writing to `package.json` (see above). This is causing lots of issues when updating projects, especially when bulk-updating many projects and is prone to errors.

### Additional information:

 - `node -v` prints:  `v16.15.0`
 - Windows, macOS, or Linux?: Windows 10

**Comment by zkochan** [comment]
I couldn't reproduce this issue. I have created a new project and added the dependency:

```
pnpm init
pnpm add gsap@^3.10.4
```

The spec in the package.json contains the caret

**Comment by bencresty** [comment]
> The spec in the package.json contains the caret

No, it doesn't. At least not on Windows. When running the exact same commands as you do this is the resulting package.json, so clearly NO caret:

``` json
{
  "name": "test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "gsap": "3.10.4"
  }
}

```

I tried both the old build in Windows command line as well as the latest Microsoft Terminal app using Powershell. Both fail.

**Comment by mcmxcdev** [comment]
I cannot reproduce this either (on macOS). I ran the exact commands that Zoltan did. 

Make sure to check that you didn't mess around with https://pnpm.io/npmrc#save-prefix setting.

**Comment by bencresty** [comment]
@mcmxcdev  Please stop telling it cannot be reproduced if you're not using Windows. It doesn't help neither of us.

It's not working here, I do exactly what's supposed to AND the same commands as written above AND with several Windows terminals and it does NOT work. I also never changed any setting.

For the record: NPM does work just fine with all the same commands.

Please just fix this for Windows.

**Comment by zkochan** [comment]
I was able to reproduce this on Windows, using the command prompt. On Windows using PowerShell it seems like it works fine.

**Comment by bencresty** [comment]
> I was able to reproduce this on Windows, using the command prompt. On Windows using PowerShell it seems like it works fine.

I just copied the exact line of the changed code you wrote (so with the added double quotes), but here it still doesn't work in neither of the prompts. Still the same result. Also in NPM it works without the need to add double quotes around the version number.

I'm using Windows 10, 21H2 (OS Build 19044.1766)

**Comment by zkochan** [comment]
ok, so it actually passes the argument to both npm and pnpm as gsap@3.10.4, when you run `gsap@^3.10.4`

The only reason this works as you expect it to work with npm, is because npm adds the care even if you run `npm install gsap@3.10.4`

**Comment by bencresty** [comment]
@zkochan I removed my post you reacted to as I did an extra check. But see you reacted to my post. Please see below for the new reaction:

> I was able to reproduce this on Windows, using the command prompt. On Windows using PowerShell it seems like it works fine.
> 
> This works as expected with the command prompt as well:
> 
> ```
> pnpm install gsap@"^3.10.4"
> ```

@zkochan Thanks for testing. And good that you could reproduce it at least in the prompt.

USING PNPM:
- I just tested your change line, so now with double quotes, in the command prompt, and then it is working. So it only works with the added double quotes.
- But when using it with Powershell, which I always use, either via the Microsoft Terminal App, or VS Code ** or via nodeJS scripts!!**, so the one needed, unfortunately it still doesn't work. Not with and not without the double quotes. I also tested it directly into the PowerShell app: still the same.

USING NPM:
- NPM Works fine with the caret in Windows Command Prompt, **whithout the double quotes**
- NPM Works fine with the caret in PowerShell, whithout the double quotes

WINDOWS
I'm using Windows 10, 21H2 (OS Build 19044.1766) and never changed any setting for prompts, neither did I change any setting on PNPM.




> ok, so it actually passes the argument to both npm and pnpm as gsap@3.10.4, when you run `gsap@^3.10.4`
> 
> The only reason this works as you expect it to work with npm, is because npm adds the caret even if you run `npm install gsap@3.10.4`

Yes, when running `npm install gsap@3.10.4` in Powershell (so NPM) it installs WITH adding the caret to package.json. Which I'm all for as that's a nice default. Because why would we want to stay to a static version when we update? The only time we want to stay at the same version always while running an update, would be if there's something wrong with the next minor version. So that would not be the norm.

I'm all for changing the behaviour of PNPM to be the same as NPM; so always add that caret when running `pnpm install gsap@3.10.4`.

BTW If you guys are planning to change this, which I obviously hope, than please also change this situation which is different in PNPM compared to NPM and IMO npm does better:

When installing like this: `npm install gsap@3` NPM adds to the package.json file: `gsap@^3.10.4`, so the latest minor version of v3 of gsap.

But when running `pnpm install gsap@3` (so PNPM), PNPM adds to the package.json file just a `3` as version number, which is pretty odd to me. We would expect the same result as the NPM output, so instead of adding `gsap@3` (which NPM does), we should get `gsap@^3.10.4`, just like with NPM. That does makes sense, because we want the last minor version of version 3 and want it to update to the latest minor version when we run an update.

Hope you guys could change this too!!

Thanks in advance and for your fast responses!

**Comment by aakash14goplani** [comment]
I am facing the same issue with my windows machine and pnpm

**Comment by LoranRendel** [comment]
The same problem on Windows.

When I use `pnpm i bootstrap@^5.3.0-alpha3`, the caret is missing.
Now I made my own npm package and have the same behaviour.

When I use `pnpm up` the caret disappears from my package but not the others.

**Comment by shinokada** [comment]
I found that in order to work, the version in package.json must have a caret.
For example, running `pnpm i -D svelte@"^5.12.0" won't work if you have `"svelte": "5.11.0",` in your package.json.
But if you have `"svelte": "^5.11.0",`, the command works.

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm add semver` writes `^7.8.5` to package.json, keeping the caret.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
